core/runtime.rs

```diff
@@ -316,38 +327,68 @@ impl Drop for JsRuntime {
   }
 }
 
-impl JsRuntime {
+impl <const FOR_SNAPSHOT: bool> JsRuntimeImpl<FOR_SNAPSHOT> {
   const STATE_DATA_OFFSET: u32 = 0;
   const MODULE_MAP_DATA_OFFSET: u32 = 1;
 
-  /// Only constructor, configuration is done through `options`.
-  pub fn new(mut options: RuntimeOptions) -> Self {
+  pub fn new_for_snapshot(mut options: RuntimeOptions, runtime_snapshot_options: RuntimeSnapshotOptions) -> JsRuntimeImpl<true> {
     let v8_platform = options.v8_platform.take();
 
     static DENO_INIT: Once = Once::new();
-    DENO_INIT.call_once(move || v8_init(v8_platform, options.will_snapshot));
+    DENO_INIT.call_once(move || v8_init(v8_platform, true));
 
-    // Add builtins extension
-    // TODO(bartlomieju): remove this in favor of `SnapshotOptions`.
-    let has_startup_snapshot = options.startup_snapshot.is_some();
-    if !has_startup_snapshot {
-      options
-        .extensions
-        .insert(0, crate::ops_builtin::core::init_ops_and_esm());
-    } else {
-      options
-        .extensions
-        .insert(0, crate::ops_builtin::core::init_ops());
+    let snapshot_options = snapshot_util::SnapshotOptions::new_from(
+      options.startup_snapshot.take(),
+      true,
+    );
+
+    // TODO(mmastrac): We should not be printing here
+    #[cfg(feature = "include_js_files_for_snapshotting")]
+    for source in options
+      .extensions
+      .iter()
+      .flat_map(|e| vec![e.get_esm_sources(), e.get_js_sources()])
+      .flatten()
+      .flatten()
+    {
+      use crate::ExtensionFileSourceCode;
+      if let ExtensionFileSourceCode::LoadedFromFsDuringSnapshot(path) =
+        &source.code
+      {
+        println!("cargo:rerun-if-changed={}", path.display())
+      }
     }
```

Comment:
We should expose `pub(crate) JsRuntime::extensions() -> &Vec<Extensions>`, have `create_snapshot()` return this list of files and print them in `{cli,runtime}/build.rs`.

Reply:
Completed -- much cleaner.